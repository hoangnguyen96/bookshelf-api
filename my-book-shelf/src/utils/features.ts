import {
  FieldErrors,
  FieldValues,
  Path,
  UseFormClearErrors,
} from "react-hook-form";
import { v4 as uuidv4 } from "uuid";

/**
 * Clear error message when the users typing
 */
export const clearErrorOnChange = <T extends FieldValues>(
  fieldName: Path<T>,
  errors: FieldErrors<T>,
  clearErrorFunc: UseFormClearErrors<T>
): void => {
  errors[fieldName]?.message && clearErrorFunc(fieldName);
};

export const generateSevenDigitUUID = () => {
  const uuid = uuidv4().replace(/\D/g, "");
  return uuid.slice(0, 7);
};
