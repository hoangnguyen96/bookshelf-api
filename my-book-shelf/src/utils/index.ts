export * from "./book";
export * from "./date";
export * from "./error";
export * from "./user";
export * from "./validators";
