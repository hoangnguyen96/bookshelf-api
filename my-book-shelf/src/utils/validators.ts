import { ERROR_MESSAGES } from "@app/constants";

export const isRequired = (value: string | null | undefined): boolean =>
  !!value;

export const isValidFormat = (value = "", pattern: RegExp): boolean =>
  pattern.test(value);

/**
 * @param requiredFields [] The required fields on form
 * @param dirtyFields [] The fields, which the users touched and fill data on
 * @param errors {} The errors fields
 * NOTE: If the user touched and fill data for the fields, which defined on array requiredFields and without errors message
 *  ==> The button should enable.
 * When the button enable AND user focusing on the last element
 * the UX: hit `enter` on the last field to submit form should work
 */
export const isEnableSubmitButton = (
  requiredFields: string[],
  dirtyFields: string[],
  errors: Record<string, unknown>
): boolean => {
  const isMatchAllRequiredFields: boolean = requiredFields.every((field) =>
    dirtyFields.includes(field)
  );

  return isMatchAllRequiredFields && errors && !Object.keys(errors).length;
};

export const validateRequired = (value: string | null): string | true =>
  isRequired(value?.trim()) || ERROR_MESSAGES.FIELD_REQUIRED;

export const validateRegExpFormat = (
  value: string,
  pattern: RegExp,
  ariaLabel: string
): string | true =>
  isValidFormat(value.trim(), pattern) || ERROR_MESSAGES.FORMAT(ariaLabel);

export const validateConfirmPassword = (
  confirmPassword: string,
  password: string
) => confirmPassword === password || ERROR_MESSAGES.CONFIRM_PASSWORD;
