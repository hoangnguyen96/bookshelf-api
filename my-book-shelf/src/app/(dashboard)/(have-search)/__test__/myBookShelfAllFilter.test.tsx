import { act, fireEvent, render } from "@testing-library/react";
import MyBookShelfByParams from "../my-book-shelf/(main)/[...slug]/page";
import { useSession } from "next-auth/react";
import { DATA_BOOKS, DATA_USER } from "@app/app/__mocks__/data";
import * as utils from "@app/utils";
import { getAllBook, getBookById, getUserById } from "@app/api";

jest.mock("next-auth/react", () => ({
  useSession: jest.fn(),
}));

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

jest.mock("@app/api", () => ({
  getAllBook: jest.fn(),
  getUserById: jest.fn(),
  getBookById: jest.fn(),
  updateBookById: jest.fn(),
}));

jest.mock("@app/utils", () => ({
  ...jest.requireActual("@app/utils"),
  filterBooksOnShelf: jest.fn(),
  filterBooksOnShelfByParams: jest.fn(),
}));

describe("My Book Shelf All Search", () => {
  const mockBooks = DATA_BOOKS.filter((item) =>
    DATA_USER[0].shelfBooks.includes(item.id)
  );

  const mockBooksByParams = mockBooks.filter((book) =>
    book.title.includes("on")
  );

  (useSession as jest.Mock).mockReturnValue({
    data: {
      user: {
        isAdmin: true,
        email: "admin@gmail.com",
        id: "3733403",
        name: "admin",
        image: "https://i.ibb.co/RHMqQGr/man-1.png",
      },
    },
    status: "authenticated",
  });

  beforeEach(() => {
    jest.clearAllMocks();
    jest.spyOn(utils, "filterBooksOnShelf").mockReturnValue(mockBooks);
    jest
      .spyOn(utils, "filterBooksOnShelfByParams")
      .mockReturnValue(mockBooksByParams);

    (getAllBook as jest.Mock).mockReturnValue({
      data: DATA_BOOKS,
    });
    (getUserById as jest.Mock).mockReturnValue({
      data: DATA_USER[0],
    });
    (getBookById as jest.Mock).mockReturnValue({
      data: mockBooks[0],
    });
  });

  it("Should render correctly snapshot", async () => {
    await act(async () => {
      const { asFragment } = render(
        <MyBookShelfByParams params={{ slug: ["title", "on"] }} />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("Should handle return book", async () => {
    const { findAllByTestId } = render(
      <MyBookShelfByParams params={{ slug: ["title", "on"] }} />
    );

    const buttons = await findAllByTestId("return-book");

    fireEvent.click(buttons[0]);

    expect(getBookById).toHaveBeenCalledWith(mockBooks[0].id.toString());
  });
});
