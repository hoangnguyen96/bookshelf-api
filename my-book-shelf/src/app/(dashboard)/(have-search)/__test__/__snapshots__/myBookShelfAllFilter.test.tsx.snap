// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Book Shelf All Search Should render correctly snapshot 1`] = `<DocumentFragment />`;
